### PR TITLE
server: rework Gossip bootstrap sequence

### LINKED ISSUE #13027: gossip: Restarting a cluster with new hostnames/IPs doesn't work in many cases
This was found while testing out the restore functionality from a 10-node cluster backup onto a new set of 10 nodes. When starting up the cluster with three different configurations of `--join` parameters on the 10 nodes, something different happened each time:

1. With `--join=<node-1>` for nodes 2-10 and `--join=""` for node 1: none of the nodes initialized properly. They all got stuck waiting to be connected to the cluster via gossip.
1. With `--join=<node-(i-1)>` for nodes i=2 to i=10, and `--join=""` for node 1: nodes 4-10 initialized properly. Nodes 1-3 got stuck waiting to be connected to the cluster via gossip.
1. With `--join=<comma-separated-list-of-all-nodes>` for all the nodes in the cluster, including node 1: the cluster started up properly.

Because all of the nodes were new, it's worth noting that none of the addresses saved from the gossip network before the backup were usable on the new nodes. Thus, while all the nodes had a store full of data, they didn't know where the other nodes were other than what they got in the `--join` parameter.

While the details vary a little (see below), the core issue in both of the first two cases is that nodes don't serving incoming connections until they have initialized themselves, where "initialized themselves" effectively means that the node is connected enough to know the cluster ID. This means that only nodes with a copy of the first range can properly bootstrap themselves if they're unable to connect to another node, because they're the only ones that will add the cluster ID to their local gossip infostore.

In the first case above, because node 1 didn't have a copy of range 1, it is unable to initialize itself. Because it doesn't get initialized, it doesn't start accepting incoming connections, and thus it doesn't get the cluster ID even though it has incoming connections from 3 nodes that do have the cluster ID (the 3 that have a copy of range 1). Thus the cluster never gets going -- 3 nodes are up and running in some sense, but don't know about each other so they can't do anything.

In the second case above, nodes 4-10 work because node 4 is the lowest node ID that had a copy of range 1. Thus, it initializes itself and accepts the incoming connection from node 5, which allowed node 5 to initialize, which allowed node 6 to initialize, etc. Nodes 1-3 are stuck forever because they have no outgoing gossip connection to any of nodes 4-10.

In the third case above, all nodes work because they all attempt to connect to at least one node that has a copy of range 1.

-------------

It would be conceivable to say that the current state of things is ok, and that you should specify all the nodes in all the join flags (or at the very least ensure that it's a fully-connected directed graph). That seems a bit prescriptive and error-prone, though. If a node was able to learn from incoming gossip connections during startup, pretty much any configuration of `--join` flags would work.

@mjibson @spencerkimball

### LINKED ISSUE #18027: acceptance: local cluster is blocked on problems with empty join flag (#13027)
See https://github.com/cockroachdb/cockroach/pull/18007.

`TestMixedVersion` runs with three nodes as otherwise it's flaky: the node first restarted (i.e. the one without join flags) never manages to connect to gossip. See the comment within that test.

One theory is that the other three nodes manage to connect among each other sufficiently so that the don't feel that establishing new connections is helpful. In turn, the first node remains alone and unable to do anything about it. The theory is easy to test by running the test with a version of cockroach that always connects to everything it can.

It repros easily using `time go test -v -tags acceptance ./pkg/acceptance/ -run Mixed  -show-logs -count 100 -timeout 1h` within only a few iters (after making the test 4nodes, of course).

### PR DESCRIPTION
**Pull request by tbg**
This is a WIP because I want to add the script from #19427, which reproduces
this problem with 100% accuracy, as an acceptance test.

----

A long time ago, in #5441, we changed the server startup sequence so that we
would only open the main listener at the very end of the startup sequence. The
impetus was that by opening the listener early, various services could receive
requests when they weren't actually ready, which resulted in unexpected errors
during tests.

What we missed at the time was that this isn't actually a good idea because it
also robs the node of the ability to accept incoming Gossip connections. As a
result, if a node didn't manage to talk to Gossip peers (perhaps it was started
without a join flag and had nothing reasonable persisted), it would never get
its Gossip connected and would never leave `(*Node).start` which has to wait for
Gossip.

These deadlocks worsened the impact of some of our other insufficiently
streamlined aspects of the startup sequence, namely the choice of `--join`
flags, and were generally hard to debug especially since the HTTP endpoints, and
thus many debugging facilities, weren't available.

This commit reworks the sequence so that

- the HTTP endpoint opens early (so that the debug endpoints are accessible)
- opens the main cmux listener early
    - but delays opening the Postgres sublistener
    - but fails all non-Gossip and non-Ping gRPC traffic retryably

Fixes #18027.
Fixes #13027.

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM (I haven't read line-by-line but I assume it's basically just code movement)

Some more important context is that in 1.0, the server didn't open its port until it was ready to serve, so TCP-level health checks were sufficient. With the introduction of the `init` command in 1.1, a not-yet-initialized server will open its port early (which means that higher-level health checks are required for determining readiness). Now that we do that, we might as well also have initialized-but-not-connected servers do the same. 

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/server.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEUJk9mjpkBT0KkHXu:-KxEUJk9mjpkBT0KkHXv:bdm1u12) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e533905830421c082ce6b839a9f96a374a686ef/pkg/server/server.go#L200)):*
> ```Go
> 
> 	{
> 		var earlyRPCWhiteList = map[string]struct{}{
> ```

Can we unify this with the init server? It would be nice to only have one version of this interceptor. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxETA_7wLO5oS9GWDxs:b-fyu2xs)*

**Comment by a-robinson** [comment]
:lgtm:, but ditto on having glossed over some of the code movement. Have you verified that this fixes situations like your gist in #19427?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxEqE8-02H9oX3qlvJf:b-r0v0u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
> 19427

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxEqSkoCpR_SXeJo147:b-chszww)*

**Comment by tschottdorf** [comment]
> Have you verified that this fixes situations like your gist in #19427?

Yep, by now I finished translating that into an acceptance test, too.

> Some more important context is that in 1.0, the server didn't open its port until it was ready to serve, so TCP-level health checks were sufficient. With the introduction of the init command in 1.1, a not-yet-initialized server will open its port early (which means that higher-level health checks are required for determining readiness). Now that we do that, we might as well also have initialized-but-not-connected servers do the same.

Anything actionable in this PR? Should the health endpoint be opened later?

PTAL at the new test.

---

Review status: 1 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/server/server.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEUJk9mjpkBT0KkHXu:-KxJ-8cyMxp1RrqtRUQ-:bey20oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e533905830421c082ce6b839a9f96a374a686ef/pkg/server/server.go#L200)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we unify this with the init server? It would be nice to only have one version of this interceptor. 
</blockquote></details>

Good idea, though I'll file this for now: https://github.com/cockroachdb/cockroach/issues/19523

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxIqmqu6B1i-yboZuSf:b-4wrawl) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> 19427
</blockquote></details>

Happy little accident?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxIpDa8a_HCDoVtU2m8:b-pwalof)*

**Comment by bdarnell** [comment]
No, I wasn't suggesting any changes to this PR, but just thought that comment might be worth preserving somewhere (maybe in the commit message?)

---

Reviewed 1 of 1 files at r2, 4 of 5 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/gossip_peerings_test.go, line 189 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ56yfX6Vi3vLVhOLM:-KxJ56yfX6Vi3vLVhOLN:b-15caap) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L189)):*
> ```Go
> 
> // This test verifies that when the first node is restarted and has no valid
> // join flags or persisted Gossip bootstrap records, it will still be able to
> ```

Why doesn't the first node have persisted gossip bootstrap records?

---

*[pkg/acceptance/gossip_peerings_test.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ4m76QVbCHBEH-uUi:-KxJ4m76QVbCHBEH-uUj:b-qcuu2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L217)):*
> ```Go
> 
> 	lc := c.(*localcluster.LocalCluster)
> 	// Upgrade the first node's binary to match the other nodes (i.e. the testing binary).
> ```

Copy-pasta comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxJ3mcWKiRbAxvui__F:b-bb99e1)*

**Comment by tschottdorf** [comment]
Ack, done. TFTRs!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/gossip_peerings_test.go, line 189 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ56yfX6Vi3vLVhOLM:-KxJ7QtFOO_YcDoJSw-F:blm6a8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L189)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why doesn't the first node have persisted gossip bootstrap records?
</blockquote></details>

The test sets up things so that the first node's bootstrap info is all outdated. Added comment.

---

*[pkg/acceptance/gossip_peerings_test.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ4m76QVbCHBEH-uUi:-KxJ7a3DsAxatAdE9mYS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L217)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Copy-pasta comment?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxJ8Tcje2IbPmIGwOza:b-swe23a)*

**Comment by a-robinson** [comment]
Review status: 2 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxJ7T_92KiXHBIkLtyZ:b-rgzsmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Happy little accident?
</blockquote></details>

"Happy little accident" could be a good replacement for the shame cube

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493)*

File: pkg/rpc/context.go [hunk 1]
```diff
@@ -131,15 +140,64 @@ func NewServer(ctx *Context) *grpc.Server {
...
+		// well. The otgrpc package has no such facility, but there's also this:
+		//
+		// https://github.com/grpc-ecosystem/go-grpc-middleware/tree/master/tracing/opentracing
+	}
+
+	// TODO(tschottdorf): when setting up the interceptors below, could make the
+	// functions a wee bit more performant by hoisting some of the nil checks
+	// out. Doubt measurements can tell the difference though.
+
+	if interceptor != nil {
```
Comment (tamird) [inline pkg/rpc/context.go]: Some of this code might benefit from https://godoc.org/github.com/grpc-ecosystem/go-grpc-middleware ?

**Comment by tschottdorf** [comment]
@bdarnell what are your thoughts on cherry-picking this to 1.1.2? I think it can wait, though the issue that is being fixed is sure to cause a lot of problems when it occurs.

**Comment by bdarnell** [comment]
I'm reluctant to make this change in a patch release; this is a potentially disruptive change (see #19559. It could also require changes to LB configurations). However, this depends on the availability of other solutions, since the problem being solved is pretty severe. If you follow best practices in setting join flags, how likely are you to run into this problem? Does it only occur if there are IP changes, or could it be more common?

**Comment by a-robinson** [comment]
To me, this seems too big/risky of a change to be patched into 1.1.

As far as I'm aware, the only situations we've constructed in which it becomes a problem are when IPs or ports change, which shouldn't be common in real deployments.

**Comment by tschottdorf** [comment]
I'm ok letting this bake until 2.0. I don't think you run into this much outside of the intentionally exotic things our acceptance tests do.

**Pull request merged**